Week4/Assignment2/README.md

```diff
@@ -0,0 +1,645 @@
+|-----------------------|-------------------------------------------|------------------------------------------|-----------------------------------------|
+| **Definition**        | Dependency is injected via the constructor.| Dependency is injected directly into a field. | Dependency is injected via a setter method. |
+| **Advantages**        | - Ensures all dependencies are injected when the object is created. <br> - Promotes immutability. | - Simple and concise. <br> - Less boilerplate code. | - Allows for optional dependencies. <br> - Can be used to change dependencies after object creation. |
+| **Disadvantages**     | Can lead to constructor bloat if there are many dependencies. | - Breaks encapsulation by making fields accessible to the framework. <br> - Difficult to perform unit testing. | Can leave the object in an inconsistent state if not called. |
+| **Usage**             | `@Autowired` on the constructor.          | `@Autowired` directly on the field.      | `@Autowired` on the setter method.      |
+| **Best Practice**     | Recommended for mandatory dependencies and better testability. | Not recommended due to encapsulation issues and testability concerns. | Recommended for optional dependencies.  |
+
+## Conclusion
+
+### Which one is better?
+
+- **Constructor Injection** is generally considered the best practice for mandatory dependencies because it ensures that the dependency is available when the object is created, promoting immutability and making the code easier to test.
+
+### Which one is recommended?
+
+- **Constructor Injection** is recommended for most cases, especially for mandatory dependencies.
+- **Setter Injection** can be useful for optional dependencies or when you need to change the dependency after object creation.
+- **Field Injection** is typically not recommended due to its negative impact on encapsulation and testability, although it can be used for simplicity in some cases.
+
+# Number 3: Circular Dependency Injection
+![alt text](image.png)
+
+## What is Circular Dependency Injection?
+
+Circular dependency injection occurs when two or more beans in a dependency injection framework (like Spring) depend on each other. For example, if `ClassA` depends on `ClassB`, and `ClassB` depends on `ClassA`, a circular dependency is created.
+
+## Example of Circular Dependency
+
+Imagine we have the following classes:
+
+```java
+public class ClassA {
+    private ClassB classB;
+
+    @Autowired
+    public ClassA(@Lazy ClassB classB) {
+        this.classB = classB;
+    }
+}
+
+public class ClassB {
+    private ClassA classA;
+
+    @Autowired
+    public ClassB(@Lazy ClassA classA) {
+        this.classA = classA;
+    }
+}
+```
+
+In this scenario, Spring cannot resolve the dependencies because it gets stuck in an infinite loop trying to create an instance of `ClassA` and `ClassB`.
```

Comment:
The example above is similar to the solution for avoiding errors due to circular dependencies using the @Lazy annotation. To illustrate a scenario that could lead to a BeanCurrentlyInCreationException, it would be better if we removed the @Lazy annotation from the constructor.